Lib/io.py

```diff
@@ -97,3 +99,55 @@ class TextIOBase(_io._TextIOBase, IOBase):
     pass
 else:
     RawIOBase.register(_WindowsConsoleIO)
+
+#
+# Static Typing Support
+#
+
+GenericAlias = type(list[int])
+
+
+class Reader(metaclass=abc.ABCMeta):
+    """Protocol for simple I/O reader instances.
+
+    This protocol only supports blocking I/O.
+    """
+
+    __slots__ = ()
+
+    @abc.abstractmethod
+    def read(self, size=..., /):
```

Comment:
Sorry if this has been discussed before, but I'm unsure on the runtime use of `size=...` (I didn't notice this earlier in my documentation review, sorry).

Almost every other `read(size)` method I can find has a default of either `None` or `-1`. I also can't find another method in the stdlib with a default of `...` (outside of the recently-added protocols in `wsgiref.types`).

Would it be better to have `size=-1`, to indicate that the method takes an `int`? I'm not sure how much we want typeshed-like practices to leak into the standard library.

```suggestion
    def read(self, size=-1, /):
```

A

Reply:
This is not really something you can do in a protocol. I also wouldn't want to mandate that implementors have to use a default of `-1`, because – as you said – some implementations use `None`.

Reply:
Right, but this isn't a protocol -- it's an ABC, which do have defaults -- see e.g. `collections.abc.Generator`. All of the `read()` methods in `io` are documented as having `read(size=-1, /)`, and given these ABCs are going into `io`, I think we should be consistent with that interface, or have good reason to diverge from it (& document why).

* https://docs.python.org/3/library/io.html#io.RawIOBase.read
* https://docs.python.org/3/library/io.html#io.BufferedIOBase.read
* https://docs.python.org/3/library/io.html#io.BufferedReader.read
* https://docs.python.org/3/library/io.html#io.TextIOBase.read

Reply:
It's supposed to be a protocol, not an ABC. (Notwithstanding the fact that all protocols are ABCs.) It's just a protocol in the implementation for performance reasons. And using `-1` as a default would give users the very wrong impression that they can use `read(-1)` when that may or may not actually be supported.

Reply:
> using `-1` as a default would give users the very wrong impression that they can use `read(-1)` when that may or may not actually be supported.

From the documentation of `io.RawIOBase.read()`:

> Read up to *size* bytes from the object and return them. As a convenience, if *size* is unspecified or -1, all bytes until EOF are returned.

I would expect the `io.Reader.read()` ABC/protocol to have this same guarantee, for a 'properly' implemented `read()` method (according to the `io` expecations). In the proposed documentation, we say:

> Read data from the input stream and return it. If *size* is specified, it should be an integer, and at most size items (bytes/characters) will be read.

This forbids `None` (good!), but is silent on what happens should *size* be omitted. I still think we should use `-1` instead of `...`, but at the very least we should include in the documentation the contract for what happens when `read()` is called with no arguments.

Reply:
I disagree. `-1` is the default for *some* implementations, but the protocol should not make any default mandatory.

Reply:
I'm still not sure I follow, though -- currently the ABC/protocol has the mandatory default of `size=...`, which is always invalid and not the correct type. `-1` is valid for all interfaces specified by the `io` documentation, which is where this new type is being added.

I ran the following with `mypy --strict` and it passed, so I don't think type checkers care about default values (and as discussed, the type forbids using non-integer types):

```py
import abc

class Reader(metaclass=abc.ABCMeta):
    __slots__ = ()

    @abc.abstractmethod
    def read(self, size: int = -1, /) -> bytes: pass

class CustomReader(Reader):
    def read(self, size: int = -1, /) -> bytes:
        return b''

class CustomReaderZero(Reader):
    def read(self, size: int = 0, /) -> bytes:
        return b''

assert issubclass(CustomReader, Reader)
assert issubclass(CustomReaderZero, Reader)
assert isinstance(CustomReader(), Reader)
assert isinstance(CustomReaderZero(), Reader)

def reader_func(r: Reader) -> None:
    r.read()

reader_func(CustomReader())
reader_func(CustomReaderZero())
```

Reply:
I agree with Sebastian here; we should use `...` because the protocol need not mandate any particular default.

> -1 is valid for all interfaces specified by the io documentation

The protocol should also match other file-like classes defined elsewhere in the stdlib or even in third-party libraries. When defining a protocol it's often useful to be permissive, so that all objects that are intended to match the protocol actually match it.